### DISCUSSION TITLE
[BP#74] Ne jamais faire de SELECT * FROM

### DISCUSSION THREAD

**Discussion by yannkozon** [cnumr/best-practices#102] [Bonnes pratiques] [answered]
J'aurais une tendance à reformuler la bonne pratique pour la rendre valide dans tous les cas d'utilisations.
"Utiliser uniquement les champs utilisés lors d'une requête à une base de données"

**Comment by ACTLEM** [comment]
Effectivement, en phase avec cette remarque.

**Comment by fzaninotto** [comment]
Je me permets de challenger cette pratique. 

Ce que l'on économise en ne faisant pas de SELECT * est un peu de transfert réseau entre la BDD et le serveur web, trafic dont le coût écologique tend vers zéro. Quand aux impacts de performance, ils dépendent grandement du modèle de données et de la présence de colonnes de taille non limitée. Dans la plupart des cas, il seront nuls. 

Le fait de ne pas utiliser de SELECT * interdit également l'utilisation d'ORMs, qui permettent pourtant de faire des optimisations dans le requêtage des bases de données. Cela peut donc être contre-productif. 

A mon avis cette pratique a un impact très irrégulier, et l'ampleur de cet impact est non prouvé. Je pense qu'il faudrait la supprimer.

**Reply by rlemaire** [reply]
Les ORM ne font pas de select * (en tout cas ceux que je connais).

**Reply by tbroyer** [reply]
Sur l'impact d'un wildcard: quand on fait un "suggest" ou une liste déroulante ou autre (bref, dès qu'on liste des données avec un nombre restreint de données affichées: un identifiant et/ou un libellé, parfois quelques données complémentaires), utiliser un wildcard alors que la table a un grand nombre de colonnes, y compris des colonnes potentiellement volumineuses (une "description longue" par exemple ; ou le contenu d'un article/une page pour un CMS), aura vraisemblablement un impact non-négligeable.

Sur les ORMs, je n'ai jamais vu un ORM "faire des optimisations dans le requêtage des bases de données", bien au contraire. Je suis preneur d'exemples si tu en as.

**Comment by rlemaire** [comment]
Je propose de renommer la pratique vers quelque chose du genre "Optimiser les requêtes aux bases de données" (wording très imparfait).

Les select * sont rares, surtout à l'époque des ORM et effectivement souvent on aurait de toute façon récupéré toutes les colonnes. 

Par contre se poser systématiquement la question d'optimiser les requêtes est intéressant, et c'est bien le message associé à cette pratique.

Aussi dans le cas où un ORM est utilisé, il y a une tendance à récupérer tous les objets en entiers et pas seulement les informations nécessaires pour l'affichage ou le traitement, ce qui est une piste d'optimisation dans certains cas.

**Reply by devcrafting** [reply]
Cela me fait pensé à CQRS avec l'idée qu'il est préférable d'utiliser des modèles différents pour "l'écriture" (C=command, un modèle) et les lectures (Q=query, plusieurs modèles par cas d'utilisation sans X couches classiques et X frameworks surchargeant le serveur...). Et effet de bord c'est généralement plus performant...reste qu'il faut s'y former car comme tout ça peut produire le pire...

**Reply by devcrafting** [reply]
J'avais en tête de proposer une BP autour de CQRS, mais plusieurs BP existantes peuvent y être lié (j'en ai parlé aussi dans la discussion sur les données en cache #78)

**Comment by tbroyer** [comment]
Les ORM sont probablement une des pires idées qu'on ait pu avoir. En ecoconception on devrait probablement privilégier des query-builders (ou des requêtes SQL directement) justement par la maîtrise qu'ils apportent sur "ce qui se passe" entre l'application et la base de données, et dès qu'on est au niveau SQL utiliser un wildcard dans la projection fait partie des mauvaises pratiques unanimement reconnues (c'est utile quand on "joue" avec sa base de données dans un outil ad-hoc, mais à proscrire dans le code applicatif). Faire le mapping des résultats vers des objets n'est pas ce que j'appelle un ORM.

Donc si on veut aller plus loin, on devrait aussi indiquer d'éviter les ORMs qui ne laissent pas la possibilité de restreindre les champs récupérés (il y en a qui le permettent, souvent à grand renfort de configuration), et faire en sorte de toujours sélectionner uniquement les champs dont on a besoin (qu'on soit au niveau SQL/query-builder ou via un ORM, et puisqu'un ORM permet trop facilement de ne pas se poser la question et récupérer trop de données –sans parler parfois des données liées–, on devrait privilégier des solutions qui nous obligent à faire un choix explicit, donc plutôt SQL/query-builder). On serait donc là sur un choix de conception/architecture technique, et pas simplement une optimisation technique ou une bonne pratique "d'écriture de code".

Je pense par contre qu'on devrait laisser cette BP centrée sur la quantité de données échangées avec le serveur de base de données. L'optimisation des requêtes au sens large pourrait faire partie d'une autre BP, mais on serait là sur de l'optimisation technique de performance la plupart du temps.

**Reply by rlemaire** [reply]
"Faire le mapping des résultats vers des objets n'est pas ce que j'appelle un ORM." ORM veut dire Object Relationnal Mapping. Je ne comprends pas trop du coup? Peut-être qu'on ne parle pas tout à fait de la même chose ?

Les ORM sont des outils, qui peuvent bien ou mal utilisés. Ralentir les applications n'était certainement pas l'idée de Gavin King. Ils sont des avantages très intéressants, comme de pouvoir utiliser le même code pour une base en mémoire pour les tests, une base libre comme mysql en hors prod et un outil payant comme Oracle en prod. Il faut les apprendre et les comprendre, et comprendre ce qu'ils font. On peut restreindre les champs récupérés, il suffit de faire des requêtes, en HQL / JQL par exemple. On peut aussi utiliser du SQL spécifique si on souhaite optimiser une requête spéciale. Quand on parle d'optimisation, la règle de 80/20 s'applique souvent. Dans beaucoup d'applications, optimiser quelques échanges (requêtes) sera souvent suffisant.

Je ne sais pas si seules la quantité de données compte, il faut aussi compter avec l'impact du serveur de bases de données.

**Reply by fzaninotto** [reply]
Je lis beaucoup d'opinions mais peu de faits. Où sont les preuves qu'un SELECT * dégrade l'impact d'une application ? Où sont les preuves qu'un ORM dégrade l'impact d'une application ? 

Cette discussion est emblématique d'une bonne partie des pratiques du guide, qui sont des préférences, souvent discutables, et loin d'être universelles, mais pas des faits avérés.

En l'absence de preuves, je pense qu'il vaut mieux se dispenser d'unclure une pratique pour ne pas discréditer l'ensemble (ce que des soit-disant bonnes pratiques comme "utiliser des simples quotes plutôt que des guillements" ont eu pour effet dans mon équipe).

**Reply by tbroyer** [reply]
> "Faire le mapping des résultats vers des objets n'est pas ce que j'appelle un ORM." ORM veut dire Object Relationnal Mapping. Je ne comprends pas trop du coup? Peut-être qu'on ne parle pas tout à fait de la même chose ?

Je fais la différence par exemple, dans le monde Java, entre jOOQ (hors "records"), Jdbi ou SQLDelight d'une part et JPA/Hibernate ou Spring Data d'autre part. JPA et Spring Data favorisent le "select *" (ou l'équivalent: "select <tous les champs connus de l'entité>") avec une approche basée sur une représentation objet du domaine, indépendante de l'usage (par défaut, je ne me pose pas la question "de quels champs ai-je besoin?", je les récupère tous dans mon objet "métier" ; je peux le faire mais ça demande un travail *supplémentaire*), quand jOOQ, Jdbi ou SQLDelight sont "SQL-centric" et obligent à écrire explicitement le comportement voulu (même si avec jOOQ on a des raccourcis pour faire un "select <tous les champs connus de l'entité>", mais ils restent quand même explicites).

Mais la plus grande différence entre ces approches vient des "données liées", avec JPA qui va faire du lazy-loading des données liées sur l'appel d'un getter par exemple, et un développeur va pouvoir (et on le constate en pratique) "oublier" la façon dont les données sont récupérées et se contenter de travailler avec ses objets, et on se retrouve avec 150 requêtes SQL (quand c'est pas plus) pour répondre à une unique requête HTTP à cause de ces appels incontrôlés, parce que cachés par cette couche d'abstraction et particulièrement son fonctionnement par défaut (@fzaninotto c'est du vécu, plus d'une fois, et jamais plaisant à debugger). Sans compter que `@ManyToOne` et `@OneToOne` sont _eager_ par défaut, donc vont faire une jointure ou une seconde requête et récupérer des données même si on ne les utilise pas (et en cascade s'il vous plait, sinon ce ne serait pas drôle).
Bien évidemment qu'on peut "bien utiliser" ces outils, tout comme on peut assez facilement "mal utiliser" ceux que j'ai placé dans la catégorie opposée, mais c'est leur comportement par défaut, leur "getting started" qui me dérange principalement, il faut faire un travail supplémentaire (qui peut très bien ne pas être anodin) pour avoid un comportement "optimum", en définissant par exemple des [entity graphs](https://vladmihalcea.com/jpa-entity-graph/), soit en les prévoyant sur l'entité (une sorte "d'action at a distance"), soit en utilisant une API complexe et facilement rebutante, mais dans tous les cas un travail supplémentaire et absolument pas intuitif.
Quand Vlad Mihalcea [conseille de toujours vérifier](https://vladmihalcea.com/hibernate-facts-always-check-criteria-api-sql-queries/) ce que fait Hibernate, en d'autres termes ne jamais lui faire confiance, pour moi c'est un signe d'une erreur de conception à la base (YMMV). 

> Les ORM sont des outils, qui peuvent bien ou mal utilisés. Ralentir les applications n'était certainement pas l'idée de Gavin King. Ils sont des avantages très intéressants, comme de pouvoir utiliser le même code pour une base en mémoire pour les tests, une base libre comme mysql en hors prod et un outil payant comme Oracle en prod.

jOOQ aussi, sans les autres inconvénients.
(je ne m'étendrais pas sur la pertinence d'utiliser des bases différentes entre tests / hors prod / prod, ce n'est pas le sujet, mais si on utilise la même base partout, le polyglottisme n'est alors plus un argument, et deviendrait presque un contre-argument puisqu'il ajoute une phase de traduction, un traitement, dont on ne peut se passer par design ; “The fastest code is the code which does not run.” – Robert Galanakis)

> Je ne sais pas si seules la quantité de données compte, il faut aussi compter avec l'impact du serveur de bases de données.

Oui mais on parle alors de "calculs", et c'est autre chose ; il y a probablement plusieurs façons d'écrire une requête qui ramènera les mêmes données mais avec des performances totalement différentes. Il s'agit d'optimisation, alors que "les données dont j'ai besoin" (et ne pas en ramener trop) est plus proche de (et en partie dictée par) la user-story que de l'implémentation technique finalement.

**Reply by devcrafting** [reply]
Totalement en phase avec l'explication de @tbroyer sur les ORMs

Sur les données vs les calculs, je suis plutôt en phase avec @rlemaire : il faut également considérer le calcul qui impacte le dimensionnement des serveurs de base de données (et l'inflation fréquente avec du scaling vertical et donc une presseion au remplacement de matériel - même dans des contextes d'infra avec virtualisation/containairisation)

Pour la remarque de @fzaninotto je propose une autre réponse pour avoir un fil dédié 😉

**Comment by ACTLEM** [comment]
Pour alimenter la discussion, un retour terrain:
- Service web à destination de comptables magasin qui a pour but de générer le rapport comptable de la veille (au sein de l'UF: Contrôler et valider les encaissements magasin)
- Dans un magasin avec 10 000 transactions quotidiennes (oui c'est un beau magasin ^^) le rapport mettait 2h pour s'afficher, autant dire inutilisable pour les comptables.
- Le serveur BDD a été boosté pour passer à 45mins la génération du rapport. Les impacts ont été multipliées côté serveur et restaient excessifs côté terminal utilisateur (poste du comptable allumé en attendant de pouvoir avoir le rapport et de travailler effectivement).
- Optimisation de la requête sur l'ancienne configuration du serveur BDD: 12 mins! (soit divisé le temps d'attente du comptable divisé par 10)

Je vous rassure on ne s'est pas arrêté là, d'autres améliorations (fonctionnelles et techniques) ont été réalisées pour arriver à 2 mins.

Tout cela pour dire que je pense qu'il faut garder une BP spécifique sur les requêtes SQL. Le `select * from` est effectivement trop restrictif  (dans mon cas ce n'était pas cela le problème) et je pense que @yannkozon a raison de proposer une amélioration. 
A l'inverse "Optimiser les requêtes SQL" est à mon avis trop générique et difficile à vérifier factuellement, mais faute de mieux cela peut être mise en avant via un niveau de mise en oeuvre complexe (demande une expertise DBA).
La règle de validation pourrait s'appuyer l'analyse des 10 requêtes les plus utilisées et coûteuses (Explain plan).

**Reply by arnolem** [reply]
Il faut clairement une ou plusieurs règles sur l'accès aux données en SQL.
Mais on est d'accord que dans ton cas, ce type de problème n'arrivera jamais à cause d'un `SELECT *`.

**Reply by devcrafting** [reply]
Oui le SELECT * me semble aussi anecdotique, les discussions ici me semblent bien plus intéressantes

**Comment by arnolem** [comment]
Le coût d'un `SELECT *` est négligeable dans la plupart des cas. 
En tout cas, il y a de nombreuses améliorations possibles au niveau SQL mais celle-ci n'est du tout la priorité à mon sens.

J'aurais mettre en TOP 1 : **Utiliser convenablement des index sur vos tables**  pour économiser les ressources serveurs.
Voici une vidéo ultra intéressante sur les index : https://www.youtube.com/watch?v=-C2js48fAQU

**Reply by devcrafting** [reply]
Il va certainement falloir plusieurs BP pour tout ce qui tourne autour de l'accès aux données 😅

**Comment by ACTLEM** [comment]
Pour infos, un papier sur une BDD Oracle pour choix uniquement des colonnes nécessaires (donc comparaison du select * from): https://tanelpoder.com/posts/reasons-why-select-star-is-bad-for-sql-performance/

C'est assez logique, mais ça permet d'avoir quelques métriques.

**Reply by ACTLEM** [reply]
Voici les tests que j'ai réalisé sur une BDD PostgresQL (700 000 objets avec une douzaine de champs, et récupération de 10 000 éléments):

`SELECT *`:

![postgres_star](https://user-images.githubusercontent.com/15525949/139727111-7e8931a5-2834-4719-bb3a-7c6d403476b0.png)


`SELECT id, label, category` (donc 3 champs):

![postgres_field](https://user-images.githubusercontent.com/15525949/139727139-54b37cf4-fc31-43e7-a9a7-4b6a40713931.png)


Sur la partie mémoire, il n'y a pas d'effet. Sur le CPU, l'usage total maximum est identique, la différence réside surtout dans le temps d'exécution qui peut amener de la contention en cas de charge importante mais de manière nominale le levier paraît peu important.
Côté serveur applicatif (Spring Boot java pour mon test), la différence est peu notable.
  
Les impacts sont donc en réalité peu visibles d'après mes tests sur une base PostgresQL. Comme le dit @arnolem, ce n'est pas du tout la priorité n°1 quand il s'agit d'optimiser côté BDD même si ne pas ramener des champs qui ne sont pas utilisés est du bon sens (cela aura une incidence en cas de traitements lourds sur la mémoire, ce qui est rarement le cas dans une archi web).
Je suis donc plutôt pour trouver d'autres bonnes pratiques axées sur la BDD. Le challenge est de trouver une bonne pratique qui a un réel effet et a une règle de validation assez simple.

**Reply by tbroyer** [reply]
Voir aussi https://blog.jooq.org/10-more-common-mistakes-java-developers-make-when-writing-sql/#:~:text=Returning%20too%20many%20columns

> This mistake is quite frequent and can lead to very bad effects both in your database’s execution plan and in your Java application. Let’s look at the second effect first: **Bad effects on the Java application**: If you’re selecting * (star) or a “default” set of 50 columns, which you’re reusing among various DAOs, you’re transferring lots of data from the database into a JDBC ResultSet. Even if you’re not reading the data from the ResultSet, it has been transferred over the wire and loaded into your memory by the JDBC driver. That’s quite a waste of IO and memory if you _know_ that you’re only going to need 2-3 of those columns. This was obvious, but beware also of… **Bad effects on the database execution plan**: These effects may actually be much worse than the effects on the Java application. Sophisticated databases perform a lot of SQL transformation when calculating the best execution plan for your query. It may well be that some parts of your query can be “transformed away”, knowing that they won’t contribute to the projection (SELECT clause) or to the filtering predicates.

**Comment by devcrafting** [comment]
Pour faire un fil dédié à la remarque de @fzaninotto https://github.com/cnumr/best-practices/discussions/102#discussioncomment-1544856

> Je lis beaucoup d'opinions mais peu de faits. Où sont les preuves qu'un SELECT * dégrade l'impact d'une application ? Où sont les preuves qu'un ORM dégrade l'impact d'une application ?

Pour le SELECT * c'est assez rapide à démontrer je pense : faire un bout de code en n'importe quel langage qui accède à une table (avec un certain nombre de colonnes de différents types) avec SELECT */SELECT colonnes utiles, et mesurer le temps CPU BDD + bout de code.

Concernant les ORMs, c'est un vaste sujet d'architecture à mon sens. Utiliser un ORM pour des cas d'utilisation d'affichage induit bien souvent des requêtes totalement inadaptées (cf exemples de @tbroyer), car remontant beaucoup trop de données, en utilisant des strategies de fetching génériques pas toujours bien senties (en mode plutôt black box, sauf pour les rares devs voulant bien ouvrir le capot). Il ne me semble pas très compliqué d'imaginer faire un même projet avec une architecture basée sur un ORM et une autre architecture, pour comparer l'usage des ressources. Je pourrais même tenter de faire cela sur cette base de code : https://github.com/devcrafting/architecture-styles

Les ORMs ont les inconvénients de leur atout : en encapsulant beaucoup et avec une promesse de simplicité (typique du "Getting Started"/Tuto), de mon expérience (consultant sur plusieurs projets), l'effet est désastreux à l'usage à l'échelle tant sur les performances que sur la sobriété numérique. Sans compter qu'au final il faut des ultra-spécialistes (très rares) pour arriver à maintenir une base de code bien faite avec un ORM...alors que des architectures plus simples techniquement pourraient être plus accessibles et plus maintenables (moins de couplage). Il est intéressant de lire sur l'architecture de StackOverflow (faut que je retrouve les sources).

NB: ressenti personnel depuis 2004 : ce qui est fou c'est qu'on est passé d'un monde du dev sans architecture (caricature : tout en vrac dans le contrôleur/la page web) à une architecture ultra normée et pilotée par des frameworks très contraignants...on a (un peu) zappé l'étape apprendre des concepts d'architecture et la qualité logicielle par tous les devs (même si ça vient un peu) 

> Cette discussion est emblématique d'une bonne partie des pratiques du guide, qui sont des préférences, souvent discutables, et loin d'être universelles, mais pas des faits avérés.

Je n'aurai peut-être pas dit "une bonne partie", mais il faut à coup sûr sourcer et argumenter ce qu'on avance pour ne pas être trop péremptoire.

> En l'absence de preuves, je pense qu'il vaut mieux se dispenser d'unclure une pratique pour ne pas discréditer l'ensemble (ce que des soit-disant bonnes pratiques comme "utiliser des simples quotes plutôt que des guillements" ont eu pour effet dans mon équipe).

Je suis bien d'accord avec ce constat. C'est tout l'enjeu de cette nouvelle version et le moteur de ma participation 🥳

**Comment by ACTLEM** [comment]
Plutôt pour la suppression du "Select * from" mais aussi pour avoir une BP minimum sur le SQL

@yannkozon , c'est possible de créer un point visio (15-30mins) pour discuter avec les personnes qui seraient intéressées afin de proposer concrètement une bonne pratique stp?

**Reply by devcrafting** [reply]
J'en suis si possible (mais ça va être chaud avec le planning serré)

**Comment by ACTLEM** [comment, marked as answer]
Issue pour suppression: https://github.com/cnumr/best-practices/issues/250

Si vous avez des idées pour des BP autour des BDD/SQL, n'hésitez pas à ajouter (d'ici le 19/11 pour la V4).
